libpod/container_internal.go

```diff
@@ -1042,6 +1042,91 @@ func (c *Container) cniHosts() string {
 	return hosts
 }
 
+// Run networksetup hooks
+func (c *Container) runNetworkSetupHooks(ctx context.Context) error {
+	allHooks := make(map[string][]spec.Hook)
+	if c.runtime.config.Engine.HooksDir == nil {
+		if rootless.IsRootless() {
+			return nil
+		}
+		for _, hDir := range []string{hooks.DefaultDir, hooks.OverrideDir} {
+			manager, err := hooks.New(ctx, []string{hDir}, []string{"networksetup"})
+			if err != nil {
+				if os.IsNotExist(err) {
+					continue
+				}
+				return err
+			}
+			ociHooks, err := manager.Hooks(c.Spec(), c.Spec().Annotations, len(c.config.UserVolumes) > 0)
+			if err != nil {
+				return err
+			}
+			if len(ociHooks) > 0 || c.Spec().Hooks != nil {
+				logrus.Warnf("implicit hook directories are deprecated; set --ociHooks-dir=%q explicitly to continue to load ociHooks from this directory", hDir)
+			}
+			for i, hook := range ociHooks {
+				allHooks[i] = hook
+			}
+		}
+	} else {
+		manager, err := hooks.New(ctx, c.runtime.config.Engine.HooksDir, []string{"networksetup"})
+		if err != nil {
+			return err
+		}
+
+		allHooks, err = manager.Hooks(c.Spec(), c.Spec().Annotations, len(c.config.UserVolumes) > 0)
+		if err != nil {
+			return err
+		}
+	}
+
+	state, err := json.Marshal(spec.State{
+		Version:     spec.Version,
+		ID:          c.ID(),
+		Status:      "networksetup",
+		Bundle:      c.bundlePath(),
+		Annotations: c.config.Spec.Annotations,
+	})
+	if err != nil {
+		return err
+	}
+
+	netnsPath := ""
+	if !c.config.PostConfigureNetNS && c.state.NetNS != nil {
```

Comment:
It seems like this hook stage isn't actually guaranteed to run after network init, given that `PostConfigureNetNS` exists. That seems a little questionable - I suppose the limitation could be documented, but it's still counterintuitive.

Reply:
That was the question that I had in mind... To ensure the hook fires after the network init in any case, should I make the PR to OCI spec and rework this PR to pass the hook similar to `createContainer` one?

Usually, what chances are that an OCI specification proposal gets approved in the reasonable timeframe?

Reply:
By "OCI spec" I meant OCI spec + libcontainer + runc + (maybe) crun PRs. Thinking about kata and runsc I do not use...

Reply:
I don't see any example where `PostConfigureNetNS` is set. Can you point the one, please?

Reply:
Main use of it is with user namespaces in root containers - we need to configure the network after the user namespace is created.

Reply:
> I don't see any example of Podman invocation where `PostConfigureNetNS` is set. Can you point the one, please?

OK, I see from https://github.com/containers/podman/pull/3310 that the primary use-case for `PostConfigureNetNS` is VM OCI runtimes like Kata. Still, the question: if slirp4netns starts routing packets, how different is behavior with Kata from the behavior with rootless Podman? I want to have a clear understanding of the issue before I can make the limitation documented or overcome it.